Convert all references to lowercase (#1619)
conan-io/docs: creating_packages/define_abi_compatibility.rst

be able to differentiate the packages built with ``intel`` with the ones built by ``Visual Studio`` because both will have the same package ID,
and that is not always desirable.

.. _problem_of_dependencies:

Dependency Issues
-----------------

Let's define a simple scenario whereby there are two packages: ``my_other_lib/2.0`` and ``my_lib/1.0`` which depends on
``my_other_lib/2.0``. Let's assume that their recipes and binaries have already been created and uploaded to a Conan remote.

> Now, a new release for ``my_other_lib/2.1`` is released with an improved recipe and new binaries. The ``my_lib/1.0`` is modified and is required to be upgraded to ``MyOtherLib/2.1``.

Reviewer: ```suggestion
Now, a new release for ``my_other_lib/2.1`` is released with an improved recipe and new binaries. The ``my_lib/1.0`` is modified and is required to be upgraded to ``my_other_lib/2.1``.
```

    from conans.model.version import Version

    class PkgConan(ConanFile):
        name = "my_lib"
        version = "1.0"
        settings = "os", "compiler", "build_type", "arch"
        requires = "my_other_lib/2.0@lasote/stable"

        def package_id(self):

>             myotherlib = self.info.requires["MyOtherLib"]

Reviewer: ```suggestion
            myotherlib = self.info.requires["my_other_lib"]
```

And below this line, there are other `MyOtherLib` that we should change

  .. code-block:: text

>     MyLib/1.2.3@user/testing       => MyLib/1.Y.Z

Reviewer: ```suggestion
    my_lib/1.2.3@user/testing       => my_lib/1.Y.Z
```

And above this, there is another code-block that requires a change

    my_other_lib/2.3.4@user/testing  =>

  So the direct dependencies are mapped to the major version only. Changing its channel, or using version

>   ``MyLib/1.4.5`` will still produce ``MyLib/1.Y.Z`` and thus the same package-id.

Reviewer: ```suggestion
  ``my_lib/1.4.5`` will still produce ``my_lib/1.Y.Z`` and thus the same package-id.
```

  .. code-block:: python

    def package_id(self):
      self.info.requires["my_other_lib"].semver_mode()

  This results in:

  .. code-block:: text

>     MyLib/1.2.3@user/testing       => MyLib/1.Y.Z

Reviewer: ```suggestion
    my_lib/1.2.3@user/testing       => my_lib/1.Y.Z
```

>     MyOtherLib/2.3.4@user/testing  => my_other_lib/2.Y.Z

Reviewer: ```suggestion
    my_other_lib/2.3.4@user/testing  => my_other_lib/2.Y.Z
```

  .. code-block:: python

      def package_id(self):
          self.info.requires["my_other_lib"].base_mode()

- ``full_version_mode()``: Any changes to the version of the required dependency changes the package ID.

  .. code-block:: python

    def package_id(self):
      self.info.requires["my_other_lib"].full_version_mode()

  .. code-block:: text

    my_other_lib/1.3.4-a4+b3@user/testing  => my_other_lib/1.3.4-a4+b3   

- ``full_recipe_mode()``: Any change in the reference of the requirement (user & channel too) changes the package ID.

  .. code-block:: python

    def package_id(self):
        self.info.requires["my_other_lib"].full_recipe_mode()

  This keeps the whole dependency reference, except the package-id of the dependency.

  .. code-block:: text

    my_other_lib/1.3.4-a4+b3@user/testing  => my_other_lib/1.3.4-a4+b3@user/testing   

- ``full_package_mode()``: Any change in the required version, user, channel or package ID changes the package ID.

  .. code-block:: python

    def package_id(self):

>         self.info.requires["MyOtherLib"].full_package_mode()

Reviewer: ```suggestion
        self.info.requires["my_other_lib"].full_package_mode()
```

- ``app/1.0`` now depends on ```pkgc/1.0`` and ``pkga/2.0``

.. image:: /images/conan-full_transitive_package_id.png
    :width: 100 %
    :align: center

With the default behavior, the header only ``pkgb`` is isolating ``pkgc`` from the upstream changes effects. The package-id ``PIDC1`` we
get for ``pkgc/1.0`` is exactly the same when depending on ``pkga/1.0`` and ``pkga/2.0``.

If we want to have the ``full_version_mode`` to be fully transitive, irrespective of the local package-id modes of the packages,
we can configure it in the :ref:`conan_conf` section. To summarize, you can activate the ``general.full_transitive_package_id``
configuration (``$ conan config set general.full_transitive_package_id=1``).

If we do this, then ``pkgc/1.0`` will compute 2 different package-ids, one for ``pkga/1.0`` (``PIDC1``) and the other to link with
``pkga/2.0`` (``PIDC2``).

Library Types: Shared, Static, Header-only
++++++++++++++++++++++++++++++++++++++++++

Let's see some examples, corresponding to common scenarios:

- ``my_lib/1.0`` is a shared library that links with a static library ``my_other_lib/2.0`` package.

>   When a new ``MyOtherLib/2.1`` version is released: Do I need to create a new binary for

Reviewer: ```suggestion
  When a new ``my_other_lib/2.1`` version is released: Do I need to create a new binary for
```

      def package_id(self):
          # Any change in the my_other_lib version, user or
          # channel or Package ID will affect our package ID
          self.info.requires["my_other_lib"].full_package_mode()

- ``my_lib/1.0`` is a shared library, requiring another shared library ``my_other_lib/2.0`` package.
  When a new ``my_other_lib/2.1`` version is released: Do I need to create a new binary for
  ``my_lib/1.0`` to link with it?

  It depends. If the public headers have not changed at all, it is not necessary. Actually it might
  be necessary to consider transitive dependencies that are shared among the public headers, how
  they are linked and if they cross the frontiers of the API, it might also lead to
  incompatibilities. If the public headers have changed, it would depend on what changes and how are
  they used in ``my_lib/1.0``. Adding new methods to the public headers will have no impact, but
  changing the implementation of some functions that will be inlined when compiled from
  ``my_lib/1.0`` will definitely require re-building. For this case, it could make sense to have this configuration:

  .. code-block:: python

      def package_id(self):

>           # Any change in the MyOtherLib version, user or channel

Reviewer: ```suggestion
          # Any change in the my_other_lib version, user or channel
```

          # or Package ID will affect our package ID
          self.info.requires["my_other_lib"].full_package_mode()

>           # Or any change in the MyOtherLib version, user or

Reviewer: ```suggestion
          # Or any change in the my_other_lib version, user or
```
